### ISSUE TITLE
AsyncCacheRegion

### ISSUE THREAD

**Issue by zzzeek** [sqlalchemy/dogpile.cache#199]
adapt the approach taken by SQLAlchemy in https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/util/_concurrency_py3k.py to provide for an AsyncCacheRegion frontend.

backends will as always present a "sync" interface that uses greenlets to adapt to the async backend.  backends to start with:

* https://github.com/aio-libs/aiomcache
* https://github.com/aio-libs/aioredis

**Comment by CaselIT** [comment]
should we create a package with sqlalchemy async adapted implementation? Or ctrl-c ctrl-v is better in this case?

**Comment by zzzeek** [comment]
yeah...I think for now we would vendor it (the latter option).  dogpile.cache is python 3 only now so it can be more succinct.   if we put out "sqlalchemy/greenlet_async" then we have to support that separately, would rather not go there yet.

**Comment by antont** [comment]
Hi - am using async SQLAlchemy with FastAPI happily, and was using dogpile.cache before ported the system to async otherwise. Have been struggling to port dogpile over to async, gotten some things to work partially by hacking around in the internals of SA ORM Session, Dogpile Region etc. 

One problem is that `event.listen` is not implemented for async sessions, but am trying to hook it via the sync session etc.

Could you maybe hint at a the way to go, what would AsyncCacheRegion do? I'd have ok time to work on this on coming days, and pretty well versed with async Python in general, just not familiar with SA code from before so it's been a lot to digest.

**Comment by zzzeek** [comment]
> Hi - am using async SQLAlchemy with FastAPI happily, and was using dogpile.cache before ported the system to async otherwise. Have been struggling to port dogpile over to async, gotten some things to work partially by hacking around in the internals of SA ORM Session, Dogpile Region etc.
> 
> One problem is that `event.listen` is not implemented for async sessions, but am trying to hook it via the sync session etc.

this is specific to SQLAlchemy, so when using event.listen with the asyncsession, follow the guidelines at https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#using-events-with-the-asyncio-extension

then, when you are inside the event handler, suppose you are using aiomemcache or something like that.  no problem, you can call out to asyncio methods inside the event handler using either the connection-bound run_async() method, documented at https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#using-awaitable-only-driver-methods-in-connection-pool-and-other-events, or more directly, and we havent documented this yet, you can run async defs using [await_only](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/util/_concurrency_py3k.py#L81), just like:

```py
from sqlalchemy import event
from sqlalchemy.util import await_only

engine = create_async_engine(...)
session = AsyncSession(engine)

@event.listens_for(session.sync_session, "before_flush")
def evt(sess, context, objects):
    cache_data = await_only(my_async_cache.get("some key"))
```

> Could you maybe hint at a the way to go, what would AsyncCacheRegion do? I'd have ok time to work on this on coming days, and pretty well versed with async Python in general, just not familiar with SA code from before so it's been a lot to digest.

So this is a much bigger job and for anyone, requires a lot of cognitive work, like more than I have for the time being for sure :).  But the general idea is that it would look a lot like [AsyncSession](https://github.com/sqlalchemy/sqlalchemy/blob/3b4d62f4f72e8dfad7f38db192a6a90a8551608c/lib/sqlalchemy/ext/asyncio/session.py#L51).  That is, it has all the methods that Region has, with all the "awaitable" ones set up as "async".  then the body of each method calls out to the "sync" method on Region, making use of greenlet_spawn for all IO blocking calls.   see https://github.com/sqlalchemy/sqlalchemy/blob/3b4d62f4f72e8dfad7f38db192a6a90a8551608c/lib/sqlalchemy/ext/asyncio/session.py#L188 for an example.

**Comment by antont** [comment]
Thanks a lot - sorry for being unclear, I should have reported what had learned so far, quick comments there.

Also, the solution I was using for sync, and have tried to port to async now, is the CachingQuery from https://docs.sqlalchemy.org/en/14/_modules/examples/dogpile_caching/caching_query.html

> this is specific to SQLAlchemy, so when using event.listen with the asyncsession, follow the guidelines at https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#using-events-with-the-asyncio-extension

Understood, I actually got some success hooking `_do_orm_execute` to `asyncsession.sync_session`.

> then, when you are inside the event handler, suppose you are using aiomemcache or something like that. no problem, you can call out to asyncio methods inside the event handler using either the connection-bound run_async() method, documented at https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#using-awaitable-only-driver-methods-in-connection-pool-and-other-events, or more directly, and we havent documented this yet, you can run async defs using 

Right, the connection binding in ORM Session is where I got lost, at `_connection_for_bind`, I think because it tried to open a sync connection even though I was using AsyncSession elsewhere. It seemed that ORM Session would have async support somehow nowadays, but wasn't sure whether I was on the right track there at all.

> So this is a much bigger job and for anyone, requires a lot of cognitive work, like more than I have for the time being for sure :). But the general idea is that it would look a lot like [AsyncSession]

Makes sense - I'll give another shot in a few days, am doing some other tasks first but return to this later.

**Comment by CaselIT** [comment]
@zzzeek what's your appetite on publishing a `sqlalchemy_async_thingy` package with the async and proxy stuff so that they can be reused in dogpile (and I guess 3rd party packages)?

**Comment by zzzeek** [comment]
> @zzzeek what's your appetite on publishing a `sqlalchemy_async_thingy` package with the async and proxy stuff so that they can be reused in dogpile (and I guess 3rd party packages)?

my appetite for what I think we're talking about here, would be that we add some additional async examples into https://docs.sqlalchemy.org/en/14/orm/examples.html#module-examples.dogpile_caching and that would be it.

**Comment by CaselIT** [comment]
>  we're talking about here,

I think it would mainly be module `sqlalchemy.util._concurrency_py3k` and maybe the proxy thing we are changing here https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3771

but it's probably less work to just copy that file in dogpile, since it's not something that changes a lot.

**Comment by zzzeek** [comment]
the important thing about the dogpile example is that it's an example.  people who use it are compelled to read the source code and take on at least some degree of responsibility for it.  if we add new API then we have to maintain it.

**Comment by a76yyyy** [comment]
I'm interested in dogpile.cache, are there any examples of using dogpile.cache in an asynchronous way?

**Comment by CaselIT** [comment]
> yeah...I think for now we would vendor it (the latter option). dogpile.cache is python 3 only now so it can be more succinct. if we put out "sqlalchemy/greenlet_async" then we have to support that separately, would rather not go there yet.

Since there is now such package, maybe dogpile it could just use it

**Comment by mavwolverine** [comment]
Any update on this?
I need to cache async function. Same as #200 which was closed as duplicate of this.
If dogpile.cache currently doesn't support, what other options are available?

Thanks

**Comment by mavwolverine** [comment]
@zzzeek  As a temp hack, I was able to use async functions with the following changes.

```
+++ b/dogpile/cache/region.py
@@ -4,6 +4,7 @@ import contextlib
 import datetime
 from functools import partial
 from functools import wraps
+import inspect
 import json
 import logging
 from numbers import Number
@@ -1050,6 +1051,11 @@ class CacheRegion:
                     )
                 else:
                     created_value = creator()
+            if inspect.iscoroutinefunction(creator):
+                try:
+                    created_value.send(None)
+                except StopIteration as e:
+                    created_value = e.value
             value = self._value(created_value)

             if (
@@ -1098,6 +1104,10 @@ class CacheRegion:
             expiration_time,
             async_creator,
         ) as value:
+            if inspect.iscoroutinefunction(creator):
+                async def coro_func(value):
+                    return value
+                return coro_func(value)
             return value

     def get_or_create_multi(
```

**Comment by zzzeek** [comment]
whatever you need to do , sure!    a real async would likely be a new AsyncRegion.   may or may not benefit from using https://github.com/sqlalchemy/awaitlet to provide a thin wrapper that's easier to maintain

**Comment by israelsgalaxy** [comment]
Did anyone manage to get this working and can provide an example of how we can go about it?
